packages/assets-controllers/src/selectors/token-selectors.ts

```diff
@@ -0,0 +1,512 @@
+  marketData: TokenRatesControllerState['marketData'];
+  currencyRates: CurrencyRateState['currencyRates'];
+  accountsAssets: MultichainAssetsControllerState['accountsAssets'];
+  assetsMetadata: MultichainAssetsControllerState['assetsMetadata'];
+  balances: MultichainBalancesControllerState['balances'];
+  conversionRates: MultichainAssetsRatesControllerState['conversionRates'];
+  currentCurrency: CurrencyRateState['currentCurrency'];
+  // This is the state from AccountTrackerController. The state is different on mobile and extension
+  // accountsByChainId with a balance is the only field that both clients have in common
+  // This field could be removed once TokenBalancesController returns native balances
+  accountsByChainId: Record<
+    Hex,
+    Record<
+      Hex,
+      {
+        balance: Hex | null;
+      }
+    >
+  >;
+};
+
+const createAssetListSelector = createSelector.withTypes<AssetListState>();
+
+const selectAccountsToGroupIdMap = createAssetListSelector(
+  [(state) => state.accountTree, (state) => state.internalAccounts],
+  (accountTree, internalAccounts) => {
+    const accountsMap: Record<string, AccountGroupId> = {};
+    for (const { groups } of Object.values(accountTree.wallets)) {
+      for (const { id: accountGroupId, accounts } of Object.values(groups)) {
+        for (const accountId of accounts) {
+          const internalAccount = internalAccounts.accounts[accountId];
+
+          accountsMap[
+            // TODO: We would not need internalAccounts if evmTokens state had the accountId
+            internalAccount.type.startsWith('eip155')
+              ? internalAccount.address
+              : accountId
+          ] = accountGroupId;
+        }
+      }
+    }
+
+    return accountsMap;
+  },
+);
+
+// TODO: This selector will not be needed once the native balances are part of the evm tokens state
+const selectAllEvmAccountNativeBalances = createAssetListSelector(
+  [
+    selectAccountsToGroupIdMap,
+    (state) => state.accountsByChainId,
+    (state) => state.marketData,
+    (state) => state.currencyRates,
+    (state) => state.currentCurrency,
+  ],
+  (
+    accountsMap,
+    accountsByChainId,
+    marketData,
+    currencyRates,
+    currentCurrency,
+  ) => {
+    const groupAssets: AssetsByAccountGroup = {};
+
+    for (const [chainId, chainAccounts] of Object.entries(
+      accountsByChainId,
+    ) as [Hex, Record<Hex, { balance: Hex | null }>][]) {
+      for (const [accountAddress, accountBalance] of Object.entries(
+        chainAccounts,
+      )) {
+        const accountGroupId = accountsMap[accountAddress];
+
+        if (!groupAssets[accountGroupId]) {
+          groupAssets[accountGroupId] = {};
+        }
+
+        let groupChainAssets = groupAssets[accountGroupId][chainId];
+        if (!groupChainAssets) {
+          groupChainAssets = [];
+          groupAssets[accountGroupId][chainId] = groupChainAssets;
+        }
```

Comment:
Thoughts on using nullish coalescing assignment to shortcut setting a fallback/default value?

```suggestion
        groupAssets[accountGroupId] ??= {};
        const groupChainAssets = groupAssets[accountGroupId][chainId] ?? [];
        groupAssets[accountGroupId][chainId] ??= groupChainAssets;
```